src/app/components/shared/input/date/date.component.ts

```diff
@@ -0,0 +1,310 @@
+   */
+  public setDisabledState = (isDisabled: boolean) =>
+    (this.disabled = isDisabled);
+
+  /**
+   * Writes a new item to the element.
+   *
+   * @param value the value
+   */
+  public writeValue(value: string): void {
+    this._errors = this.validateInput(value);
+    this.setDate(this.normalizeInput(value));
+    if (this.errors.length === 0) {
+      this.onChange(this.value);
+    }
+  }
+
+  /**
+   * Method that performs synchronous validation against the provided control
+   */
+  public validate(control: AbstractControl): ValidationErrors | null {
+    return this.validateInput(control.value);
+  }
+
+  public get showValid(): boolean {
+    return isInstantiated(this.value) && this.touched && !this.showError;
+  }
+
+  public get showError(): boolean {
+    return shouldShowError(this.dirty, this.touched, this.errors);
+  }
+
+  public get errors(): string[] {
+    return getErrorMessages(this._errors, this.errorTypes);
+  }
+
+  public validateInput(value: string | null): ValidationErrors | null {
+    if (!value) {
+      // No value provided on required input
+      return this.required ? { required: true } : null;
+    } else if (value.length < this.inputLength) {
+      // Validate length of input
+      return { minLength: true };
+    } else if (value.split("-").length !== 3) {
+      // Validate delimiters exist
+      return { delimiter: true };
+    }
+
+    const segments = value.split("-");
+    const year = parseInt(segments[0], 10);
+    const month = parseInt(segments[1], 10);
+    const day = parseInt(segments[2], 10);
+
+    if (year < this.minDateStruct.year || year > this.maxDateStruct.year) {
+      // Validate year
+      return { year: true };
+    } else if (month < 1 || month > 12) {
+      // Validate month
+      return { month: true };
+    } else if (day < 1 || day > 31) {
+      // Validate day
+      return { day: true };
+    }
+
+    const currentDate = new Date(year, month - 1, day);
+    if (currentDate.getTime() < this.minDate.getTime()) {
+      // Validate date is before minimum date
+      return { minDate: true };
+    } else if (currentDate.getTime() > this.maxDate.getTime()) {
+      // Validate date is before maximum date
+      return { maxDate: true };
+    }
+
+    // Validate date is valid
+    const ngbDate = NgbDate.from({ year, month, day });
+    return this.calendar.isValid(ngbDate) ? null : { invalid: true };
+  }
+
+  public normalizeInput(value: string | null): NgbDate | null {
+    if (!value) {
+      return null;
+    }
+
+    const date = this.formatter.parse(value);
+    return date && this.calendar.isValid(NgbDate.from(date))
+      ? NgbDate.from(date)
+      : null;
+  }
+
+  public onInput(event: Event): void {
+    this.dirty = true;
+    this.touched = true;
+    this.onTouched();
+
+    const value = (event.target as HTMLInputElement).value;
+    this._errors = this.validateInput(value);
```

Comment:
if you had bound to the `dateSelect` event you would have gotten back a `NgbDate` and would have avoided having to parse the string?

Reply:
I have bound to the `dateSelect` event, this event is for when users manually input a date?

Reply:
I'm not sure on this one, but the component seems to handle this itself with very little code in https://ng-bootstrap.github.io/#/components/datepicker/examples#popup

The validators typically watch the model and validate changes, they don't need to bind to events themselves.